core/models/asset.py

```diff
@@ -482,27 +516,76 @@ def save_initial_asset_status(sender, **kwargs):
     new_condition = kwargs.get('instance')
     related_asset = new_condition.asset
-    if not new_condition.notes == \
-            related_asset.notes:
+    if not new_condition.notes == related_asset.notes:
         related_asset.notes = \
             new_condition.notes
         related_asset.save()
 
 
+@receiver(post_save, sender=AllocationHistory)
+def update_asset_status_when_allocation_changes(sender, **kwargs):
+    allocation_history = kwargs.get('instance')
+
+    if kwargs.get('created'):
+        last_status = \
+            AssetStatus.objects.filter(
+                asset=allocation_history.asset).latest('created_at')
+        if allocation_history.current_owner:
+            AssetStatus.objects.create(
+                asset=allocation_history.asset,
+                current_status=ALLOCATED,
+                previous_status=last_status.current_status
+            )
+
+
+@receiver(post_save, sender=AssetStatus)
+def update_asset_allocation_history_when_status_changes(sender, **kwargs):
+    asset_status = kwargs.get('instance')
+
+    if kwargs.get('created'):
+        try:
+            last_allocation_record = \
+                AllocationHistory.objects.filter(
+                    asset=asset_status.asset).latest('created_at')
+        except Exception:
+            return
+        if asset_status.current_status == AVAILABLE \
+                and last_allocation_record:
+            AllocationHistory.objects.create(
+                asset=asset_status.asset,
+                previous_owner=last_allocation_record.current_owner,
+                current_owner=None)
+
+
 @receiver(post_save, sender=AllocationHistory)
 def allocation_history_post_save(sender, **kwargs):
     allocation_history = kwargs.get('instance')
     asset = allocation_history.asset
-    user = allocation_history.current_owner
-    asset.assigned_to = user
+    owner = allocation_history.current_owner
+    asset.assigned_to = owner
     asset.save()
 
     if asset.assigned_to and asset.current_status == AVAILABLE:
         message = "The asset with serial number {} ".format(
             asset.serial_number) + "has been allocated to you."
-        slack.send_message(message, user=user)
+        # send slack message only to user
+        if hasattr(owner.user, 'email'):
+            slack.send_message(message, user=owner.user)
         asset_status = AssetStatus.objects.create(
             asset=asset,
             current_status=ALLOCATED
         )
         asset_status.save()
+
+
+@receiver(post_save, sender=User)
+def assetassignee_user(sender, instance, created, **kwargs):
+    if created or not hasattr(instance, 'assetassignee'):
+        AssetAssignee.objects.create(user=instance)
+    instance.assetassignee.save()
+
+
+@receiver(post_save, sender=Department)
+def assetassignee_department(sender, instance, created, **kwargs):
+    if created or not hasattr(instance, 'assetassignee'):
+        AssetAssignee.objects.create(department=instance)
```

Comment:
Add the statement below at the end of this `post_save` signal receiver to ensure a **Department** object reference on the **AssetAssignee** model is updated if it is updated.
```python
instance.assetassignee.save()
```